### PR TITLE
Fix for performance regression in ScanningWorkspaceBuilder

### LINKED ISSUE #20468: Fix performance regression in Scanning Workspace Builder
The following PR accidentally reverted some changes in ScanningWorkspaceBuilder. This issue should revert these problems.

https://github.com/mantidproject/mantid/pull/20260/files

http://builds.mantidproject.org/job/master_performancetests2/Master_branch_performance_tests/DataObjectsTest.ScanningWorkspaceBuilderTestPerformance.test_large_scanning_workspace.htm

### PR DESCRIPTION
**Pull request by ianbush**
Fix regression affecting ScanningWorkspaceBuilder performance.

**To test:**

Check all tests pass. If you want to check the regression is fixed try with and without this fix. There should be a speedup, for me it was 26 s -> 14 s.

Fixes #20468. 

*Does not need to be in the release notes.* - these changes were introduced recently.

---

#### Reviewer ####

Please comment on the following ([full description](http://www.mantidproject.org/Individual_Ticket_Testing)):

##### Code Review #####

- [ ] Is the code of an acceptable quality?
- [ ] Does the code conform to the [coding standards](http://www.mantidproject.org/Coding_Standards)?
- [ ] Are the unit tests small and test the class in isolation?
- [ ] If there are changes in the release notes then do they describe the changes appropriately?

##### Functional Tests #####

- [ ] Do changes function as described? Add comments below that describe the tests performed?
- [ ] Do the changes handle unexpected situations, e.g. bad input?
- [ ] Has the relevant (user and developer) documentation been added/updated?

Does everything look good? Mark the review as **Approve**. A member of `@mantidproject/gatekeepers` will take care of it.

### CODE REVIEW

**Comment by ianbush** [comment]
**Note** - Please leave in 3.11 for now, this should be fixed before the release.

**Review by SimonHeybrock — APPROVED** [review]
Code changes look good, ready if builds pass.

**Comment by martyngigg** [comment]
@ianbush There seem to be a few test failures

**Comment by ianbush** [comment]
The changes in this PR #20504 are a pre-requisite to fixing the tests here.

File: Framework/Beamline/src/DetectorInfo.cpp [hunk 1]
```diff
@@ -398,5 +378,56 @@ DetectorInfo::buildMergeIndices(const DetectorInfo &other) const {
   return merge;
 }
 
+// Indices returned here are the list of time indexes not to merge
+std::vector<bool>
+DetectorInfo::buildMergeSyncScanIndices(const DetectorInfo &other) const {
+  checkSizes(other);
+  std::vector<bool> merge(other.m_scanIntervals->size(), true);
+
+  for (size_t t1 = 0; t1 < other.m_scanIntervals->size(); ++t1) {
```
Comment (OwenArnold) [inline Framework/Beamline/src/DetectorInfo.cpp]: Presumably this N*M loop is necessary? Sorting scan intervals doesn't help?

Comment (SimonHeybrock) [reply Framework/Beamline/src/DetectorInfo.cpp]: This is the sync scan case, so this should rarely be larger than a few hundred each, and I suppose often one of them might be size one? @ianbush What do you think?

Comment (ianbush) [reply Framework/Beamline/src/DetectorInfo.cpp]: Yes, as this for the sync scan this should not be too expensive, especially when comparing with the loop required for `buildMergeIndices`. The most common case would be to merge a big one with one of size one (making a scanning workspace), or a big one with a smaller one (merging D2B runs into a single workspace.

**Comment by ianbush** [comment]
Just a comment, the changes @OwenArnold commented on should not have shown up in this PR, rather in PR #20504. Now the changes from https://github.com/mantidproject/mantid/pull/20504 are merged to master, and master is merged here, the correct minimal changes are shown.

**Pull request merged**